client/packages/coldchain/src/Equipment/ImportAsset/UploadTab.tsx

```diff
@@ -107,68 +195,42 @@ export const EquipmentUploadTab: FC<ImportPanel & EquipmentUploadTabProps> = ({
       setErrorMessage(t('messages.import-error'));
     }
 
-    const rows: EquipmentImportModal.ImportRow[] = [];
+    const rows: ImportRow[] = [];
     let hasErrors = false;
-    data.data.sort(function (row, row2) {
-      const number1 = getCell(row, AssetColumn.ASSET_NUMBER);
-      const number2 = getCell(row2, AssetColumn.ASSET_NUMBER);
-      return number1 < number2 ? -1 : number1 > number2 ? 1 : 0;
-    });
 
-    data.data.map((row, index) => {
-      const importRow = {} as EquipmentImportModal.ImportRow;
-      const rowErrors: string[] = [];
-      importRow.id = FnUtils.generateUUID();
-      const assetNumber = getCell(row, AssetColumn.ASSET_NUMBER);
-      if (assetNumber && assetNumber.trim() != '') {
-        importRow.assetNumber = assetNumber;
-        // check if more than one of this asset number exists if is asset number
-        if (assetNumber == rows[index - 1]?.assetNumber) {
-          rowErrors.push(t('error.duplicate-asset-number'));
-        }
-      } else {
-        rowErrors.push(
-          t('error.field-must-be-specified', { field: t('label.asset-number') })
-        );
-      }
-      const code = getCell(row, AssetColumn.CATALOGUE_ITEM_CODE);
-      if (code === undefined || code.trim() === '') {
-        rowErrors.push(
-          t('error.field-must-be-specified', {
-            field: t('label.catalogue-item-code'),
-          })
-        );
-      } else if (
-        catalogueItemData?.filter(
-          (item: { code: string | null | undefined }) => item.code == code
-        ).length === 0
-      ) {
-        rowErrors.push(
-          t('error.code-no-match', { field: t('label.catalogue-item-code') })
-        );
-      } else {
-        importRow.catalogueItemCode = code;
-      }
-      // notes, serialNumber, and installationDate aren't essential for bulk upload
-      if (getCell(row, AssetColumn.NOTES) !== undefined) {
-        importRow.notes = getCell(row, AssetColumn.NOTES);
-      }
-      if (getCell(row, AssetColumn.SERIAL_NUMBER) !== undefined) {
-        importRow.serialNumber = getCell(row, AssetColumn.SERIAL_NUMBER);
-      }
-      if (getCell(row, AssetColumn.INSTALLATION_DATE) !== undefined) {
-        importRow.installationDate = Formatter.naiveDate(
-          DateUtils.getDateOrNull(getCell(row, AssetColumn.INSTALLATION_DATE))
+    data.data.forEach((row, index) => {
+      const { addLookup, addCell, addUnique, importRow, rowErrors } =
+        getImportHelpers(row, rows, index, t);
```

Comment:
I am wondering does the rows passed to getImportHelpers update after each time it has another value pushed to it? I think I haven't quite wrapped my head around what values will be updated within nested function calls, so it isn't clear to me.
Super cool!

Reply:
it will - the variables are passed as references, so you get the updated values. 
you're right though, it isn't clear and while it works.. it's a little hard to see what's going on!

Reply:
I think it is clear! - I think it is just me not being clear on how values are passed. Thank you for explaining